### DISCUSSION TITLE
Using themes with Vite

### DISCUSSION THREAD

**Discussion by leo-petrucci** [vanilla-extract-css/vanilla-extract#1383] [Q&A]
I'm in the process of moving some apps to partially use Vanilla Extract, but I'm currently a bit stuck with Vite.

In Next.js I'm able to pass my theme:

```
const [theme, variables] = createTheme({ ...defaultTheme });
```

into `<body>` by creating a `layout.tsx` component.

However in Vite `<body>` is part of the static `index.html`
```
<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/main.tsx"></script>
  </body>
</html>
```

This then gets populated client-side with `ReactDOM.createRoot(document.getElementById("root") as HTMLElement).render`.

If I wrap my component tree in a div and make my theme available there it works in 90% of cases, but not for elements that have been rendered at the root with `React.Portal`.

Any suggestions on how to tackle this?

**Comment by leo-petrucci** [comment]
Ended up doing this, but lmk if there's a better way:
```
import { theme } from "./theme.css"

document.body.classList.add(theme)
```

**Reply by askoufis** [reply]
You could write a small vite plugin that modifies `index.html`.

Maybe something like this:
```ts
const injectThemeClassname = () => {
  return {
    name: 'html-transform',
    async transformIndexHtml(html) {
      const { theme } = await import("path/to/theme.css.ts");
      return html.replace(
        /<body>/,
        `<body class="${theme}>"`,
      );
    },
  }
}
```

Maybe there's a simpler way to do this, but this should do the job.

See https://vitejs.dev/guide/api-plugin.html#transformindexhtml for more details.
